Fix account data visibility after disabling public addressbook upload

danxuliu opened the pull request:
When the lookup server upload was disabled the scope menu was hidden in the properties of the profile information. If any property was then changed the request did not contain any scope information, so the `UsersController` automatically set them to null. However, as the controller fully overwrote the previous properties this caused the scope of the properties to end being null. In practice, this caused that if a user changed her name from the profile information when uploading to the public address book was disabled that user could not be searched either in the local instance or from trusted servers.

Moreover, the user was not able to reset the visibility from the profile information, even if uploading to the public address book was enabled again.

Although this pull request fixes the controller it does not repair any corrupted visibility, as it is not possible to know whether the user had a public, contacts or private visibility before it got corrupted. Repairing the visibility by setting by default anything above private could be seen as a privacy issue, and setting the visibility to private would hide the problem from the user if that user wants to be found by other contacts. **Update:** since #26243 any invalid scope [is treated as _local_](https://github.com/nextcloud/server/commit/278a73789e777d2ba00fb7d8b311923a590f10fc#diff-b4c86c6664250143b37dd0f655e080c6855761a5d9e8aababb1d028178f7e69dR145). Despite that, except for the original last commit (which is still needed and thus included in the `stable19` and `stable20` backports), the rest of the commits still apply.

~~Nevertheless, I was wondering if it would be worth adding a repair step to check if the visibility of any property is null and then create a notification to the user so she can fix it. Or adding a command to fix the visibility to some value (but this would be something to be explicitly done by the admin, so it would be her responsibility).~~

## How to test (main scenario)
- Log in as an admin
- Open the settings
- Open the (Administration) _Sharing_ section 
- Disable _Allow users to publish their data to a global and public address book_
- Log out as the admin
- Log in as a regular user (admins are not shown in the contacts menu)
- Open the _Personal info_ section
- Set the user name to _Test_
- In a private window, log in as another user
- Open the contacts menu and search for "Test"

### Result with this pull request

The user is found.

### Result without this pull request

The user is not found.



## How to test (corrupted visibility subscenario) - `stable19` and `stable20`
- (Generate a wrong database state by doing all the following steps in the parent commit of this branch)
- Open the settings
- Open the (Administration) _Sharing_ section 
- Disable _Allow users to publish their data to a global and public address book_
- Open the _Personal info_ section
- Set the user name to _Test_
- (Checkout this branch and reload the page)

### Result with this pull request 

Visibility of properties can be set. Visibilities with an invalid value (all of them) will show an error.

### Result without this pull request 

Visibility of properties can not be set. Even if _Allow users to publish their data to a global and public address book_ is enabled again the visibility can still not be set, as the scope is null.


## How to test (corrupted visibility subscenario)  - `stable21` and later

- (Generate a wrong database state by doing all the following steps in the parent commit of this branch)
- Open the settings
- Open the (Administration) Sharing section
- Disable Allow users to publish their data to a global and public address book
- Open the Personal info section
- Set the user name to Test
- (Checkout this branch and reload the page)

### Result with this pull request

Visibility of properties can be set. Even if there is only one value available it will be possible to show the menu to read its detailed description.

### Result without this pull request

Visibility of properties can not be set. If _Allow users to publish their data to a global and public address book_ is enabled again the visibility will be shown as _Local_ (and in the database it will be `null`).



## How to test (public visibility when public upload is disabled subscenario)
- Open the settings
- Open the _Personal info_ section
- Set the visibility of user name to _Public_
- Open the (Administration) _Sharing_ section 
- Disable _Allow users to publish their data to a global and public address book_
- Open the _Personal info_ section

### Result with this pull request 

Visibility of user name shows "Public" as selected, but disabled. Changing the visibility to a more restricted value removes the "Public" option.

### Result without this pull request 

No visibility is shown.

danxuliu committed:
Use constants from interface rather than class

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Extract default test data to a helper getter

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Change default test data to values less similar to empty values

Right now it makes no difference, but this should make future tests
clearer, specially in case of failure.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add more unit tests for setting user settings

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Respect additional user settings not covered by the controller

"AccountManager::updateUser()" wipes previous user data with whichever
user data is given (except for some adjustments, like resetting the
verified status when needed). As the controller overrode the properties
those properties would lose some of their attributes even if they are
not affected by the changes made by the controller. Now the controller
only modifies the attributes set ("value" and "scope") to prevent that.

Note that with this change the controller no longer removes the
"verified" status, but this is not a problem because, as mentioned,
"AccountManager::updateUser()" resets them when needed (for example,
when the value of the website property changes).

This change is a previous step to fix overwritting properties with null
values, and it will prevent the controller from making unexpected
changes if more attributes are added in the future.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Fix TypeError when "email" is not given in the controller request

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Fix deleting properties of user settings when not given explicitly

The controller can receive an optional subset of the properties of the
user settings; values not given are set to "null" by default. However,
those null values overwrote the previously existing values, so in
practice any value not given was deleted from the user settings. Now
only non null values overwrite the previous values.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add integration tests for searching users in contacts menu

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Guard against null phone number value

"parsePhoneNumber()" expects a string, so a TypeError would be thrown if
the phone number value is null.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Fix active scope not visible in the menu if excluded

Depending on some settings (for example, if lookup server upload is
disabled) some items can be hidden in the scope menu. However, if the
user selected an scope in the past once the settings were changed the
scope was no longer visible in the menu. Now the active scope will be
always visible in the menu, although if it is an excluded scope it will
be disabled. Selecting any other scope will then hide the excluded and
no longer active one.

When upload to the lookup server is disabled the scope menu was hidden
for display name and email in the personal information settings; now the
menu will be always shown to enable the above described behaviour.

Note that the menu will be shown even if there is a single available
scope so the user can read its description.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu commented:
/backport to stable20

danxuliu commented:
/backport to stable19

rullzer commented:
Master is Nextcloud 22 now.
If this should go into 21 it should be backported.

danxuliu commented:
/backport to stable21

danxuliu commented:
Rebased and fixed up.

rullzer reviewed (approved):
Didn't test but I trust your test writing skills

danxuliu commented:
> Didn't test but I trust your test writing skills

I hope we will not regret that :-P

danxuliu commented:
I pushed the backports to stable19 and stable20 before this was merged because they needed some adjustments but I will probably on vacation when this is merged.

danxuliu commented:
Rebased and adjusted to changes introduced in #26243. Acceptance tests failure is known and unrelated.

I dropped the last commit because it was no longer needed (it handled corrupted scope values by showing them with an error icon in the UI, although that no longer applies due to them being now returned as _local_ scope), but the rest still applies.

I have also found some issues related to #26243 (it should be possible to set federated scope even if the lookup server is disabled, and contact data should be shown in contacts menu even if scope is local), but for clarity they will be addressed in other pull requests.

MorrisJobke commented:
Conflicts 😢

blizzz reviewed (approved):
Code looks ok, tests very much appreciated :) I did some smoke testing, which was fine, too.

danxuliu commented:
> Conflicts :cry:

Rebased and resolved :-)

danxuliu commented:
> > Conflicts cry
> 
> Rebased and resolved :-)

Uh... wait, I spoke too soon, most tests fail, let me check :-)

danxuliu commented:
It seems that most of the failures were internal GitHub errors. There are some PHP CS errors, but [they seem to be in master as well](https://github.com/nextcloud/server/pull/26669). 

The Handlebars test has failed again, but in my system _build/compile-handlebars-templates.sh_ passes. So I do not know if it is a problem with GitHub or my system :shrug:

Besides that, I noticed a small issue (in some areas of the disabled menu item the cursor was a pointer, because all items in a link element (the menu trigger) [use `cursor: pointer`](https://github.com/nextcloud/server/blob/bfe87e501d83ff4dc9e5057c07464a598d951b4b/core/css/styles.scss#L63-L65)), so I fixed the CSS to handle that.

danxuliu commented:
After the last rebase everything is now green :-)
